l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+%
+% \end{function}
+%
+% \begin{function}[EXP, pTF,added = 2020-12-14]
+%   { \bitset_if_exist:N, \bitset_if_exist:c  }
+%   \begin{syntax}
+%     \cs{bitset_if_exist_p:N} \meta{bitset var}
+%     \cs{bitset_if_exist:NTF} \meta{bitset var} \Arg{true code} \Arg{false code}%
+%   \end{syntax}
+%   Tests whether the \meta{bitset var} exist.
+% \end{function}
+
+%
+% \section{Setting and clearing bits}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_set:Nn, \bitset_set:cn, \bitset_gset:Nn, \bitset_gset:cn  }
+%   \begin{syntax}
+%     \cs{bitset_set:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gset:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and sets the bit of the
+% index position it represents to 1. Index position are zero based and set from
+% right to left (so one can view the bitset as a binary number).
+% If needed the length of the bit vector is enlarged.
+% If the \meta{integer expression} evaluates to a negative number, nothing happens.
+% \end{function}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_clear:Nn, \bitset_clear:cn, \bitset_gclear:Nn, \bitset_clear:cn }
+%   \begin{syntax}
+%     \cs{bitset_clear:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gclear:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and clears the bit of the
+% index position it represents (sets it to 0).
+% The index is zero based. If the \meta{integer expression} evaluates to a
+% negative number or is larger than the current length of the bit vector
+% nothing happens. If the leading (left most) bit is cleared,
+% zeros are not trimmed but stay in the bit vector.
+% \end{function}
+%
+% \section{Using bitsets}
+%
+% \begin{function}[EXP,added = 2020-12-14]
+%   { \bitset_index:Nn, \bitset_index:cn }
+%   \begin{syntax}
+%     \cs{bitset_index:Nn}   \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and output \texttt{1} if the bit is
+% set and \texttt{0} otherwise.
+% \end{function}
+%
+% \begin{function}[EXP,added = 2020-12-13]
+%   {\bitset_to_bin:N, \bitset_to_bin:c}
+%   \begin{syntax}
+%     \cs{bitset_to_bin:N} \meta{bitset var}
+%   \end{syntax}
+% This leaves the current value of the bitset expressed as
+% a binary (string) number in the input stream.
+% If no bit has been set yet, the output is empty.
+% \end{function}
+% \begin{function}[EXP,added = 2020-12-13]
+%   {\bitset_to_arabic:N, \bitset_to_arabic:c}
+%   \begin{syntax}
+%     \cs{bitset_to_arabic:N} \meta{bitset var}
+%   \end{syntax}
+% This leaves the current value of the bitset expressed as
+% a decimal number in the input stream. If no bit has been set yet,
+% the output is zero. The function uses \cs{int_from_bin:n}, so it
+% can only be used with a bitset whose value is smaller than $2^{31}-1$.
```

Comment:
In PDF, we sometimes need bitsets with bit 31 set (represented as 32bit signed two-complement integer), so this restriction might be too restrictive. (Of course it could be argued that this is a special case and generic bitset functionality should only provide the unsigned case, but then we might need a separate function to get the PDF output)

Reply:
we will probably need a bigint module at some time to handle larger integers and then extend the functions (the bigset package uses bigintcalc).

Reply:
As this is a one-off function, we likely don't need all of bigint` to fix it: we could merge the PR (after agreement) then look to code just that conversion up 'by hand'. (I don't think there are other places `int` functions are used?)